File: Python/programmers/kakao/level3/외벽_점검.py
from itertools import permutations

def solution(n, weak, dist):
    length = len(weak)
    for i in range(length):
        weak.append(weak[i]+n)

    answer = len(dist)+1

    for start in range(length):
        for friends in list(permutations(dist, len(dist))):
            count = 1
            position = weak[start] + friends[count-1]

            for i in range(start, start+length):
                if position < weak[i]:
                    count += 1
                    if count > len(dist):
                        break
                    position = weak[i] + friends[count-1]
        
            answer = min(count, answer)

    return -1 if answer > len(dist) else answer


n = 12
weak = [1, 5, 6, 10]
dist = [1, 2, 3, 4]
print(solution(n, weak, dist)) # 2

n = 12
weak = [1, 3, 4, 9, 10]
dist = [3, 5, 7]
print(solution(n, weak, dist)) # 1

File: Java/programmers/kakao/level3/외벽점검.java
package CodeTest.Java.programmers.kakao.level3;

public class 외벽점검 {
    private static int[] spread;
    private static int answer;

    public int solution(int n, int[] weak, int[] dist) {
        answer = Integer.MAX_VALUE;
        
        spread = new int[weak.length*2];
        for (int i=0; i<weak.length; i++) {
            spread[i] = weak[i];
            spread[i+weak.length] = weak[i]+n;
        }

        for(int k=0; k<weak.length; k++) {
            dfs(k, 0, dist, new boolean[dist.length], new int[dist.length]);
        }

        return answer == Integer.MAX_VALUE ? -1 : answer;
    }

    private void dfs(int start, int depth, int[] dist, boolean[] visited, int[] permuted) {
        if(depth == dist.length) {
            answer = Math.min(answer, check(start, start+spread.length/2, permuted));
            return;
        }
        for(int i=0; i<dist.length; i++) {
            if(visited[i]) continue;

            visited[i] = true;
            permuted[depth] = dist[i];
            dfs(start, depth+1, dist, visited, permuted);
            visited[i] = false;
        }
    }

    private int check(int start, int end, int[] permuted) {
        int friend = 1;
        int pos = spread[start]+permuted[friend-1];

        for(int i=start; i<end; i++) {
            if(pos < spread[i]) {
                friend++;
                if (friend > permuted.length) return Integer.MAX_VALUE;
                pos = spread[i]+permuted[friend-1];
            }
        }
        return friend;
    }
}
